include/directx/d3dx12.h

```diff
@@ -4037,6 +4038,421 @@ class CD3DX12_NODE_MASK_SUBOBJECT
     D3D12_NODE_MASK m_Desc;
 };
 
+//------------------------------------------------------------------------------------------------
+// Peter: Implementation for the new CheckFeatureSupport API
+
+// Macro to set up a getter function for each entry in feature support data
+// The getter function will have the same name as the feature option name
+#define FEATURE_SUPPORT_GET(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION() const \
+{ \
+    return FEATURE.OPTION; \
+}
+
+// Macro to set up a getter function for each entry in feature support data
+// Also specifies the name for the function which can be different from the feature name
+#define FEATURE_SUPPORT_GET_NAME(RETTYPE,FEATURE,OPTION,NAME) \
+RETTYPE NAME() const \
+{\
+    return FEATURE.OPTION; \
+}
+
+// Macro to set up a getter function for feature data indexed by the graphics node ID
+// The default parameter is 0, or the first availabe graphics device node
+#define FEATURE_SUPPORT_GET_NODE_INDEXED(RETTYPE,FEATURE,OPTION) \
+RETTYPE OPTION(UINT NodeIndex=0) const \
+{\
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+// Macro to initialize a member feature data struct corresponding to the specified feature
+// Will stop the initialization process and return an error code upon failure
+#define INITIALIZE_MEMBER_DATA_CHECKED(FEATURE, MEMBER) \
+if (FAILED(m_hStatus = m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))) { \
+    return m_hStatus; \
+}
+
+// Macro to initialize a member feature data and report on failure
+// Does not stop the initialization process by itself
+#define INITIALIZE_FAILED(FEATURE, MEMBER) FAILED(m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))
+
+class CD3DX12FeatureSupport
+{
+public:
+    CD3DX12FeatureSupport()
+    : m_pDevice(nullptr)
+    , m_hStatus(E_NOTIMPL)
+    {}
+
+    CD3DX12FeatureSupport(ID3D12Device* pDevice)
+    {
+        Init(pDevice);

```

Comment:
If Init() fails, this could leave some of the data fields uninitialized.  Depending on what the desired design is, either throw an exception on error, or make sure all the member data is set to well-defined default values in case of failure.  Or we could just remove this constructor?

Reply:
At this time, only three feature checks -- shader model, root signature, and feature level -- may cause an error that makes Init() fail. All other initialization routines will not cause failure. 

I checked the current implementation. It seems shader model and root signature will only return E_INVALIDARG when it doesn't recognize the feature, which is handled in Init(). Feature level may return E_FAIL only when non of the requested feature level is supported, while Init() will pass in a list of all feature levels available in d3d12.h. 

So looks like the current init() will not fail unless 1) There is a D3D12 that supports none of the feature levels, or 2) An error other than the above can rise even if it's not written in CheckFeatureSupport or relative functions. If these cases won't happen, we probably no longer need to worry about failures.

Reply:
Updating from our Teams discussion: Some of the features may still return a specific error code not handled by Init(). New features and implementation may also introduce unexpected errors. So it's better to keep the error reporting mechanism in the class as well as in Init().

I'll follow up this discussion in the PR for the second part of feature checks. There we can discuss how to handle the exceptions with the full Init() function visible to us.